### PR TITLE
[BuildPlan] Add a way to traverse the graph and compute destinations for all modules

### PR DESCRIPTION
**Pull request by xedin**
### Motivation:

This is going to be used by sourcekit-lsp to build a graph/dictionary
of all the targets and depths they appear during package loading and
replaces the need for topological sort and `buildTriple`.

### Modifications:

- Introduces a new `depthFirstSearch` variant which traverses all of the nodes in the graph.
- Adds a new method to `BuildPlan` - `traverseModules` which goes over the graph, computes destinations and returns modules with their parents and depths they appear at (the method looks though products).

### Result:

This make it possible to switch sourcekit-lsp away from using topological sort and relying on `buildTriple`.

Paired with https://github.com/swiftlang/sourcekit-lsp/pull/1621

### CODE REVIEW

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test macOS platform

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test macOS platform

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test macOS platform

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test Linux platform

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test Windows platform

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test Windows platform

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 1]
```diff
@@ -904,6 +904,66 @@ extension BuildPlan {
...
+        ) -> Self {
+            switch product.type {
+            case .macro, .plugin:
+                .product(product, .host)
+            case .test:
+                .product(product, product.modules.contains(where: self.hasMacroDependency) ? .host : destination)
+            default:
+                .product(product, destination)
+            }
+        }
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:935]: nit: IMO initializers are more idiomatic for Swift
```suggestion
        init(
            product: ResolvedProduct,
            context destination: BuildParameters.Destination
        ) {
            switch product.type {
            case .macro, .plugin:
                self = .product(product, .host)
            case .test:
                self = .product(product, product.modules.contains(where: self.hasMacroDependency) ? .host : destination)
            default:
                self = .product(product, destination)
            }
        }
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 2]
```diff
@@ -904,6 +904,66 @@ extension BuildPlan {
...
+                .module(module, .host)
+            case .test:
+                .module(module, self.hasMacroDependency(module: module) ? .host : destination)
+            default:
+                // By default assume the destination of the context.
+                // This means that i.e. test products that reference macros
+                // would force all of their successors to be `host`
+                .module(module, destination)
+            }
+        }
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:953]: nit: IMO initializers are more idiomatic for Swift
```suggestion
        init(
            product: ResolvedProduct,
            context destination: BuildParameters.Destination
        ) {
            switch product.type {
            case .macro, .plugin:
                self = .product(product, .host)
            case .test:
                self = .product(product, product.modules.contains(where: self.hasMacroDependency) ? .host : destination)
            default:
                self = .product(product, destination)
            }
        }

        init(
            module: ResolvedModule,
            context destination: BuildParameters.Destination
        ) {
            switch module.type {
            case .macro, .plugin:
                // Macros and plugins are ways built for host
                self = .module(module, .host)
            case .test:
                self = .module(module, self.hasMacroDependency(module: module) ? .host : destination)
            default:
                // By default assume the destination of the context.
                // This means that i.e. test products that reference macros
                // would force all of their successors to be `host`
                self = .module(module, destination)
            }
        }
```

File: Sources/SourceKitLSPAPI/BuildDescription.swift [hunk 3]
```diff
@@ -142,11 +155,22 @@ public struct BuildDescription {
...
 
-    /// Returns all targets within the module graph in topological order, starting with low-level targets (that have no
-    /// dependencies).
-    public func allTargetsInTopologicalOrder(in modulesGraph: ModulesGraph) throws -> [BuildTarget] {
-        try modulesGraph.allModulesInTopologicalOrder.compactMap {
-            getBuildTarget(for: $0, in: modulesGraph)
+    public func traverseModules(
+        callback: (any BuildTarget, _ parent: (any BuildTarget)?, _ depth: Int) -> Void
+    ) {
+        // TODO: One the `targetMap` is switched over to use `IdentifiableSet<ModuleBuildDescription>`
```
Comment (MaxDesiatov) [inline Sources/SourceKitLSPAPI/BuildDescription.swift:161]: nit: a typo?
```suggestion
        // TODO: Once the `targetMap` is switched over to use `IdentifiableSet<ModuleBuildDescription>`
```

File: Tests/BuildTests/BuildPlanTests.swift [hunk 4]
```diff
@@ -6496,4 +6496,126 @@ final class BuildPlanTests: XCTestCase {
             )
         }
     }
+
+    func testTraversalOverModules() async throws {
```
Comment (MaxDesiatov) [inline Tests/BuildTests/BuildPlanTests.swift:6500]: Would you mind moving these two new test functions into a new `BuildPlanTraversalTests` test case? `BuildPlanTests` is already huge, and XCTest can only run separate `XCTestCase` subclasses in parallel, but not separate functions.

**Review by MaxDesiatov — APPROVED** [review]
LGTM, only coding style, typo, and test organization nits.

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 2 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift:953]: I see, thanks!

File: Tests/BuildTests/BuildPlanTests.swift (hunk 4 above)
Comment (xedin) [reply Tests/BuildTests/BuildPlanTests.swift:6500]: Sure!

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 2 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: Fixed!

File: Sources/SourceKitLSPAPI/BuildDescription.swift (hunk 3 above)
Comment (xedin) [reply Sources/SourceKitLSPAPI/BuildDescription.swift]: ✅

File: Tests/BuildTests/BuildPlanTests.swift (hunk 4 above)
Comment (xedin) [reply Tests/BuildTests/BuildPlanTests.swift]: ✅

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test

**Comment by xedin** [comment]
https://github.com/swiftlang/sourcekit-lsp/pull/1621
@swift-ci please test Windows platform

**Pull request merged**